src/protocol.rs

```diff
@@ -18,16 +20,63 @@ use crate::util::Hash;
 pub(crate) const MAX_MESSAGE_SIZE: usize = 1024 * 1024 * 100;
 
 /// Protocol version
-pub const VERSION: u64 = 2;
+pub const VERSION: u64 = 3;
 
-#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone, MaxSize)]
+/// Connection tokens are 32 byte values used to authorize connections.
+//
+/// These are currently unused, but exist to distinguish connection authorization
+/// from request authorization.
+pub type ConnectionToken = [u8; 32];
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, MaxSize, Clone)]
 pub(crate) struct Handshake {
     pub version: u64,
+    pub conn_token: Option<ConnectionToken>,
 }
 
 impl Handshake {
     pub fn new() -> Self {
-        Self { version: VERSION }
+        Self {
+            version: VERSION,
+            conn_token: None,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, From)]
+/// Authorization tokens are opaque byte sequences used to authorize requests.
+pub struct AuthToken {
+    bytes: Vec<u8>,
+}
+
+impl AuthToken {
+    pub fn new(bytes: Vec<u8>) -> Result<Self> {
+        ensure!(bytes.len() < MAX_MESSAGE_SIZE, "auth token is too large");
```

Comment:
This would definitely be too large, lets start with restricting a much lower number, maybe `256` bytes and see how that goes